### ISSUE TITLE
ondisk not working

### ISSUE THREAD

**Issue by luigi-asprino** [SPARQL-Anything/sparql.anything#280]
This query

```
PREFIX  fx:   <http://sparql.xyz/facade-x/ns/>

SELECT  *
WHERE
  { SERVICE <x-sparql-anything:location=https://sparql-anything.cc/example1.json,ondisk=/Users/lgu/Desktop/tmp>
      { ?s  ?p  ?o }
  }

```

gives

```
-------------
| s | p | o |
=============
-------------
```

After inspecting the code I discovered that in the case of ondisk set, the query has to indicate the graph to be queried.
We can ignore for a moment that the query is supposed to give the same result with and without the ondisk set, but this

```
PREFIX  fx:   <http://sparql.xyz/facade-x/ns/>

SELECT  *
WHERE
  { SERVICE <x-sparql-anything:location=https://sparql-anything.cc/example1.json,ondisk=/Users/lgu/Desktop/tmp>
      { GRAPH <https://sparql-anything.cc/example1.json#>
          { ?s  ?p  ?o }
      }
  }
```

raises an exception `` org.apache.jena.dboe.transaction.txn.TransactionException: Not in a transaction``.

Probably, this is because the FacadeXOpExecutor doesn't start a READ transaction before executing the operation.

**Comment by luigi-asprino** [comment]
With [this](https://github.com/SPARQL-Anything/sparql.anything/commit/fd5694f1c30971c07e0530316ff48b9cd55d562d) the OpExecutor starts a Read transaction before executing the query. I leave the issue open since I would like to avoid the user indicating the graph to be queried (as for the queries without ondisk property).

**Comment by justin2004** [comment]
oh, for on ondisk you must do `graph ?g` -- that is you can't query the union graph like you can with an in memory graph. 
i put a comment in the code about it. i'll find it.

**Comment by justin2004** [comment]
here it is:
https://github.com/SPARQL-Anything/sparql.anything/blob/v0.8-DEV/sparql-anything-model/src/main/java/com/github/sparqlanything/model/BaseFacadeXGraphBuilder.java#L143-L150

**Comment by justin2004** [comment]
i tried to make ondisk behave just like in memory but it turned out that `getUnionGraph()` isn't implemented for the DatasetGraph used when you use a TDB2 on disk.

i almost opened an apache jena bug because i remember thinking it looks like it should be implemented.

**Comment by luigi-asprino** [comment]
I tried to set the global property symUnionDefaultGraph as true as suggested [here](https://jena.apache.org/documentation/tdb/datasets.html) but didn't work.
Then, I made the [OpExecutor execute the query on the union graph (in case of ondisk set)](https://github.com/SPARQL-Anything/sparql.anything/blob/6a1b0e1208e97d31ecc1a6dee32cd379830ac2f3/sparql-anything-engine/src/main/java/com/github/sparqlanything/engine/FacadeXOpExecutor.java#L285) and it seems to be working

**Comment by luigi-asprino** [comment]
Closed via [6a1b0e1](https://github.com/SPARQL-Anything/sparql.anything/commit/6a1b0e1208e97d31ecc1a6dee32cd379830ac2f3)

**Comment by luigi-asprino** [comment]
Doesn't work when passing arguments in the BGP.

**Comment by luigi-asprino** [comment]
Closed with [29b08b1](https://github.com/SPARQL-Anything/sparql.anything/commit/29b08b16a54d488990eb9dbb6d7e2115f59efe17)
